File: YUNCHOU-E/team_python/load_data.py
# load_data.py - 加载校区节点和订单数据

import json
import csv
import numpy as np

class Order:
    def __init__(self, id, pickup_id, pickup_name, delivery_id, delivery_name, weight, S, a, b):
        self.id = id
        self.pickup_id = pickup_id
        self.pickup_name = pickup_name
        self.delivery_id = delivery_id
        self.delivery_name = delivery_name
        self.weight = weight
        self.S = S          # 就绪时间 (h)
        self.a = a          # 最早可送达 (h)
        self.b = b          # 最晚可送达 (h)

class Node:
    def __init__(self, id, name, type, x, y, lat=None, lon=None):
        self.id = id
        self.name = name
        self.type = type
        self.x = x
        self.y = y
        self.lat = lat
        self.lon = lon

class Data:
    pass

def load_data(data_dir):
    """加载校区节点和订单数据"""
    data = Data()

    # === 加载校区节点 ===
    with open(f"{data_dir}/campus_nodes.json", "r", encoding="utf-8") as f:
        raw = json.load(f)

    data.nodes = []
    for n in raw["nodes"]:
        data.nodes.append(Node(n["id"], n["name"], n["type"], n["x"], n["y"], n.get("lat"), n.get("lon")))

    data.node_ids = [n.id for n in data.nodes]
    data.node_x = [n.x for n in data.nodes]
    data.node_y = [n.y for n in data.nodes]

    # 找出起降点
    data.depots = [n for n in data.nodes if n.type == "depot"]

    # === 加载订单数据 ===
    with open(f"{data_dir}/orders.json", "r", encoding="utf-8") as f:
        raw_order = json.load(f)

    data.n_orders = len(raw_order["orders"])
    data.orders = []
    for o in raw_order["orders"]:
        data.orders.append(Order(
            id=o["id"],
            pickup_id=o["pickup_id"],
            pickup_name=o["pickup_name"],
            delivery_id=o["delivery_id"],
            delivery_name=o["delivery_name"],
            weight=o["weight_kg"],
            S=o["ready_time_h"],
            a=o["tw_early_h"],
            b=o["tw_late_h"]
        ))

    # === 加载距离矩阵（实际飞行距离）===
    dist_file = f"{data_dir}/distance_matrix.csv"
    N = len(data.nodes)
    data.dist = np.zeros((N, N))

    try:
        with open(dist_file, "r", encoding="utf-8-sig") as f:
            reader = csv.DictReader(f)
            for row in reader:
                fi = data.node_ids.index(row["from_node"].strip())
                ti = data.node_ids.index(row["to_node"].strip())
                data.dist[fi, ti] = float(row["distance_m"])
        print(f"距离矩阵: 从 distance_matrix.csv 加载（实际飞行距离）")
    except FileNotFoundError:
        # 回退：用直线距离
        for u in range(N):
            for v in range(N):
                dx = data.node_x[u] - data.node_x[v]
                dy = data.node_y[u] - data.node_y[v]
                data.dist[u, v] = np.sqrt(dx**2 + dy**2)
        print("距离矩阵: 使用直线距离（未找到 distance_matrix.csv）")

    print(f"数据加载完成: 8 架无人机, {data.n_orders} 件快递, {N} 个节点")
    return data

File: matlab/preprocessing/load_data.m
function data = load_data()
% LOAD_DATA 加载校区节点和订单数据
%   data = load_data() 返回包含节点坐标、订单信息的结构体

    base_dir = fullfile(fileparts(mfilename('fullpath')), '..', '..', 'data');

    % === 加载校区节点 ===
    raw = jsondecode(fileread(fullfile(base_dir, 'campus_nodes.json')));

    data.nodes = struct();
    for i = 1:length(raw.nodes)
        data.nodes(i).id = raw.nodes(i).id;
        data.nodes(i).name = raw.nodes(i).name;
        data.nodes(i).type = raw.nodes(i).type;
        data.nodes(i).x = raw.nodes(i).x;
        data.nodes(i).y = raw.nodes(i).y;
    end

    % 找出起降点
    data.depots = {};
    for i = 1:length(data.nodes)
        if strcmp(data.nodes(i).type, 'depot')
            data.depots{end+1} = data.nodes(i);
        end
    end

    % === 加载订单数据 ===
    raw_order = jsondecode(fileread(fullfile(base_dir, 'orders.json')));

    data.n_orders = length(raw_order.orders);
    data.orders = struct();
    for j = 1:data.n_orders
        data.orders(j).id = raw_order.orders(j).id;
        data.orders(j).pickup_id = raw_order.orders(j).pickup_id;
        data.orders(j).pickup_name = raw_order.orders(j).pickup_name;
        data.orders(j).delivery_id = raw_order.orders(j).delivery_id;
        data.orders(j).delivery_name = raw_order.orders(j).delivery_name;
        data.orders(j).weight = raw_order.orders(j).weight_kg;
        data.orders(j).S = raw_order.orders(j).ready_time_h;
        data.orders(j).a = raw_order.orders(j).tw_early_h;
        data.orders(j).b = raw_order.orders(j).tw_late_h;
    end

    % === 计算节点坐标表 ===
    data.node_ids = {data.nodes.id};
    data.node_x = [data.nodes.x];
    data.node_y = [data.nodes.y];

    % === 加载距离矩阵（实际飞行距离，含绕行系数）===
    dist_file = fullfile(base_dir, 'distance_matrix.csv');
    if exist(dist_file, 'file')
        fid = fopen(dist_file, 'r');
        fgetl(fid);  % 跳过表头
        N = length(data.nodes);
        data.dist = zeros(N, N);
        while ~feof(fid)
            line = fgetl(fid);
            parts = strsplit(line, ',');
            from = strtrim(parts{1});
            to = strtrim(parts{2});
            d = str2double(parts{3});
            fi = find(strcmp(data.node_ids, from));
            ti = find(strcmp(data.node_ids, to));
            if ~isempty(fi) && ~isempty(ti)
                data.dist(fi, ti) = d;
            end
        end
        fclose(fid);
        fprintf('距离矩阵: 从 distance_matrix.csv 加载（实际飞行距离）\n');
    else
        % 回退：用直线距离
        N = length(data.nodes);
        data.dist = zeros(N, N);
        for u = 1:N
            for v = 1:N
                data.dist(u, v) = norm([data.node_x(u)-data.node_x(v), ...
                                         data.node_y(u)-data.node_y(v)]);
            end
        end
        fprintf('距离矩阵: 使用直线距离（未找到 distance_matrix.csv）\n');
    end

    fprintf('数据加载完成: %d 架无人机, %d 件快递, %d 个节点\n', ...
            raw_order.drone_count, data.n_orders, length(data.nodes));
end
